### PR TITLE
feat: add support for text-based encoding

### PR DESCRIPTION
**Pull request by mdawar**
Hey,

Currently the `Decimal` type does not support text-based encoding, for example YAML encoding will produce `{}` instead of a string representation of the decimal value.

This pull request implements the [`encoding.TextMarshaler`](https://pkg.go.dev/encoding#TextMarshaler) and [`encoding.TextUnmarshaler`](https://pkg.go.dev/encoding#TextUnmarshaler) interfaces which are recognized by many text-based encoding packages and libraries.

### CODE REVIEW

File: codec.go [hunk 1]
```diff
@@ -222,13 +236,29 @@ func (d Decimal) MarshalJSON() ([]byte, error) {
...
+	if len(data) > 2 && data[0] == '"' && data[len(data)-1] == '"' {
+		data = data[1 : len(data)-1]
+	}
+
+	return d.UnmarshalText(data)
+}
+
+// MarshalText implements the [encoding.TextMarshaler] interface.
+func (d Decimal) MarshalText() ([]byte, error) {
+	return []byte(d.String()), nil
```
Comment (quagmt) [inline codec.go:251]: I think this method should follow `MarshalJSON` implementation without quote to avoid redundant conversion from byte to string in `d.String()` and back to byte again. 
```
if !d.coef.overflow() {
        // set withQuote to false
	return d.bytesU128(true, false), nil
}

return []byte(d.stringBigInt(true)), nil

```

Comment (mdawar) [reply codec.go:251]: Yes you're totally right, I'll update it now.

Comment (quagmt) [reply codec.go]: also can you add a separate test for `UnmarshalText` only, similar to `TestUnmarshalNumber`? The reason is `UnmarshalText` usually accepts external data which can contain unexpected data. The test you added uses data from `MarshalText` and the format is correct already

Comment (mdawar) [reply codec.go]: @quagmt I did not add the same test cases for `UnmarshalNumber` as it seems that a fix is required for `UnmarshalJSON` with an empty string.

Would like me to fix it, or should it be in a separate pull request.

Comment (mdawar) [reply codec.go]: @quagmt I hope you don't mind, I've added a fix for unmarshalling empty JSON strings and added the other test cases and renamed the test to `TestUnmarshalJSON` as it's now more appropriate.

File: codec_test.go [hunk 2]
```diff
@@ -111,13 +166,7 @@ func TestUnmarshalNumber(t *testing.T) {
 
 			var test Test
 			err := json.Unmarshal([]byte(s), &test)
-			if tc.wantErr != nil {
-				require.Equal(t, tc.wantErr, err)
-				return
-			}
-
-			require.NoError(t, err)
-			require.Equal(t, strings.Trim(tc.in, `"`), test.Test.String())
```
Comment (quagmt) [inline codec_test.go:120]: why is this check removed? this ensures the data after unmarshaling matches the input and I think `TestUnmarshalText` should have a similar check

Comment (mdawar) [reply codec_test.go:120]: The `require.ErrorIs(t, err, tc.wantErr)` check will fail for `nil` values if an error is expected and it's much more idiomatic to check the error instead of checking the string value.

Anyway if you want we can revert this change.

Comment (quagmt) [reply codec_test.go:120]: I'm ok with `require.ErrorIs` but the test doesn't only check error. As I mentioned it also checks and makes sure the decimal value matches the string input

Comment (mdawar) [reply codec_test.go:120]: @quagmt Sorry I misunderstood you, I'll resolve it now, I thought you were talking about the error check.

**Review by quagmt — APPROVED** [review]
